src/core/writer.js

```diff
@@ -281,6 +283,118 @@ function updateXFA({ xfaData, xfaDatasetsRef, newRefs, xref }) {
   newRefs.push({ ref: xfaDatasetsRef, data });
 }
 
+async function getXRefTable(xrefInfo, baseOffset, newRefs, newXref, buffer) {
+  buffer.push("xref\n");
+  let refPosition = -1;
+  let lastRef = null;
+  let refCounter = -1;
+  for (const { ref, data } of newRefs) {
+    if (!lastRef || ref.num !== lastRef.num + refCounter) {
+      if (lastRef) {
+        buffer[refPosition] = `${lastRef.num} ${refCounter}\n`;
+      }
+      lastRef = ref;
+      refPosition = buffer.length;
+      buffer.push("");
+      refCounter = 1;
+    } else {
+      refCounter += 1;
+    }
```

Comment:
I'm finding this quite hard to wrap my head around, any chance that a good comment can be added to explain what's going on here?
(Or, if at all possible, perhaps the code can be simplified a little bit?)